wayback/_client.py

```diff
@@ -387,8 +402,73 @@ def reset(self):
         self.close(disable=False)
         # Re-build the standard adapters. See:
         # https://github.com/kennethreitz/requests/blob/v2.22.0/requests/sessions.py#L415-L418
-        self.mount('https://', requests.adapters.HTTPAdapter())
-        self.mount('http://', requests.adapters.HTTPAdapter())
+        if self.connection_manager:
+            self.mount('https://', PoolManagerAdapter(self.connection_manager))
+            self.mount('http://', PoolManagerAdapter(self.connection_manager))
+        else:
+            self.mount('https://', requests.adapters.HTTPAdapter())
+            self.mount('http://', requests.adapters.HTTPAdapter())
+
+
+class WaybackSession(_utils.DisableAfterCloseSession):
+    def __init__(self, max_connections=None, **kwargs):
+        self.thread_data = threading.local()
+        self.all_sessions = queue.SimpleQueue()
+        self.session_options = kwargs
+        # Create a single urllib3 poolmanager (thread-safe) to share across all
+        # the actual concrete sessions (not thread-safe).
+        if max_connections:
+            manager = PoolManager(num_pools=10, maxsize=max_connections,
+                                  block=True, strict=True)
+        else:
+            # These defaults come from requests.adapters.HTTPAdapter.
+            manager = PoolManager(num_pools=10, maxsize=10, block=False,
+                                  strict=True)
+        self.session_options['connection_manager'] = manager
+
+    def get_unsafe_session(self):
+        session = getattr(self.thread_data, 'unsafe_session', None)
+        if session is None:
+            if self._closed:
+                raise SessionClosedError('This session has already been closed '
+                                         'and cannot send new HTTP requests.')
+            session = UnsafeWaybackSession(**self.session_options)
+            self.thread_data.unsafe_session = session
+            self.all_sessions.put_nowait(session)
+        return session
+
+    def close(self):
+        while True:
+            try:
+                session = self.all_sessions.get_nowait()
+                session.close()
+            except queue.Empty:
+                break
+        super().close()
+
+    def __enter__(self):
+        concrete_session = self.get_unsafe_session()
+        return concrete_session.__enter__()
+
+    def __exit__(self, *args):
+        concrete_session = self.get_unsafe_session()
+        return concrete_session.__exit__(*args)
+
+    def request(self, *args, **kwargs):
+        concrete_session = self.get_unsafe_session()
+        return concrete_session.request(*args, **kwargs)
```

Comment:
Almost certainly there’s a better way to handle this (e.g. with magic methods, or maybe [wrapt](https://wrapt.readthedocs.io/)).